terrawrap/models/config_mover.py

```diff
@@ -0,0 +1,271 @@
+# pylint:disable=C0114,C0301
+import os
+from functools import lru_cache
+from os.path import relpath
+from pathlib import Path
+from typing import List, Union, Iterable, Any
+
+import boto3
+import git
+from botocore.exceptions import ClientError
+import hcl2
+
+from terrawrap.utils.config import find_variable_files, parse_variable_files
+from terrawrap.utils.terminal_colors import TerminalColors as Colors
+
+
+class ConfigMover:
+    """
+    Class for moving terraform configuration between directories.
+    This is done by moving .tfstate file in appropriate S3 bucket.
+    """
+
+    tf_state_file_extension = ".tfstate"
+
+    def __init__(
+        self,
+        source_directory: Path,
+        target_directory: Path,
+        tf_config_repo_root: str = "terraform-config",
+    ):
+        self.source_directory = source_directory
+        self.target_directory = target_directory
+        self.source_directory_abs = self.source_directory.absolute()
+        self.target_directory_abs = self.target_directory.absolute()
+        self.tf_config_repo_root = tf_config_repo_root
+        # we're using current path instead of the above in case they don't exist
+        self._repo = git.Repo(os.getcwd(), search_parent_directories=True)
+        self._s3 = boto3.client("s3")
+
+    @property
+    def s3_client(self):
+        """S3 client"""
+        return self._s3
+
+    @property
+    def repo(self) -> git.Repo:
+        """Git repository"""
+        return self._repo
+
+    def _to_repo_root_path(self, path: Union[str, Path]) -> Path:
+        """
+        Convert to path relative to terraform-config repository root
+        /Users/home/terraform-config/config/aws/app ->
+        terraform-config/config/aws/app
+        """
+        # [1:] removes leading slash
+        return Path(str(path).partition(self.tf_config_repo_root)[2][1:])
+
+    def _build_state_file_s3_key(self, terraform_path: Union[str, Path]) -> str:
+        """
+        Given path (absolute or relative) of terraform directory,
+        return name of this directory's state file, example:
+        Documents/terraform-config/config/aws/app/ ->
+        terraform-config/config/aws/app.tfstate
+        """
+        return f"{self.tf_config_repo_root}/{self._to_repo_root_path(terraform_path)}{self.tf_state_file_extension}"
```

Comment:
Should probably use the same function that tf itself uses
https://github.com/amplify-education/terrawrap/blob/main/terrawrap/utils/path.py#L98